spring-session-jdbc/src/integration-test/java/org/springframework/session/jdbc/AbstractJdbcIndexedSessionRepositoryITests.java

```diff
@@ -819,6 +820,33 @@ void saveNewSessionAttributeConcurrently() {
 		}
 	}
 
+	@Test
+	void savesWhenSaveModeIsOnGetAttributeAndANewAttributeIsSetAndGet() {
```

Comment:
This test would be better off in `JdbcIndexedSessionRepositoryTests`. Instead of checking the value of the attribute you can verify that it calls "INSERT INTO SPRING_SESSION_ATTRIBUTES ". The test would be similar to `saveWithSaveModeOnGetAttribute`.

Reply:
Thanks for your response!

I couldn't get the test to work inside `JdbcIndexedSessionRepositoryTests`. 

In the call to `this.repository.findById(initialSession.getId());` a NPE was thrown.
Because `List<JdbcSession> sessions = JdbcIndexedSessionRepository.this.jdbcOperations.query` `sessions` was `null`. Weird right?

Maybe because this test is using mocked `JdbcOperations`? 

---

I think testing that `INSERT INTO [...]` is executed rather than testing `getAttribute` is making the test weaker. 
As a end-user I care that after a `setAttribute` a `getAttribute` returns the correct result. I don't care how this is done internally. For this reason I think it is cleaner to test the actual API surface instead of testing what the implementation of the API does internally.

Reply:
Because the integration tests (`AbstractJdbcIndexedSessionRepositoryITests`) run on multiple implementations, 16 at the moment, we avoid adding tests for every scenario in this class.
Testing method like `setFlushMode`, `setSaveMode` etc go in `JdbcIndexedSessionRepositoryTests` class so that we can test different variations efficiently.
Additionally, the integration tests are intended to test the default `JdbcIndexedSessionRepository` that is autowired. By modifying this object we are no longer testing the default implementation. 
Also, calling `setSaveMode` at the end of the test makes the test suite fragile, since it would produce unexpected behavior if the default saveMode changed in the future.

Testing that the new attribute is inserted rather than updated is sufficient to catch this bug if we had a regression and it appeared again.
I would expect the test in `JdbcIndexedSessionRepositoryTests` to look something like this:
```
@Test
void saveWithSaveModeOnGetAttributeAndNewAttributeSetAndGet() {
	this.repository.setSaveMode(SaveMode.ON_GET_ATTRIBUTE);
	MapSession delegate = new MapSession();
	delegate.setAttribute("attribute1", (Supplier<String>) () -> "value1");
	JdbcSession session = this.repository.new JdbcSession(delegate, UUID.randomUUID().toString(), false);
	session.setAttribute("attribute2", "value2");
	session.getAttribute("attribute2");
	this.repository.save(session);
	ArgumentCaptor<BatchPreparedStatementSetter> captor = ArgumentCaptor
			.forClass(BatchPreparedStatementSetter.class);
	verify(this.jdbcOperations, times(1)).update(startsWith("INSERT INTO SPRING_SESSION_ATTRIBUTES ("),
			isA(PreparedStatementSetter.class));
	verifyNoMoreInteractions(this.jdbcOperations);
}
```